gustaf/ffd.py

```diff
@@ -1,4 +1,571 @@
+
+    
+    def _check_dimensions(self):
+        """Checks dimensions and ranges critical for a correct FFD calculation
+
+        The following things are checked:
+        is_partial:
+
+            Currently nothing since not implemented
+
+        else:
+            parametric dimension of spline is equal to geometric dimension of 
+            spline
+
+            geometric dimension of spline is the same as mesh dimension
+
+            # can currently not be checked
+            geometric range of spline is same as geometric range of mesh
+        """
+        if self._is_partial:
+            pass
+        else:
+            if not self._spline.para_dim == self._spline.dim:
+                self._logw("The parametric and geometric dimensions of the "
+                           "spline are not the same.")
+            if not self._spline.dim == self._mesh.vertices.shape[1]:
+                self._logw("The geometric dimensions of the spline and the "
+                           "dimension of the mesh are not the same.")
+            # No easy way to compare the bounds of the spline and mesh
+            # since the spline could already be deformed when initializing
+            
+            # if not self._spline.control_point_bounds == self._mesh.bounds:
+            #     self._logw("The bounds of the control "
+            #                "spline are not the same.")
+
+
+    @property
+    def mesh(self,) -> MESH_TYPES:
+        """Returns copy of current mesh. Before copying, it applies 
+        deformation.
+
+        Returns
+        -------
+        MESH_TYPES
+            Current Mesh with the deformation according to the current spline.
+        """        
+        self._deform()
+        return self._mesh.copy()
+
+    @mesh.setter
+    def mesh(self, mesh: MESH_TYPES):
+        """
+        Sets mesh. If it is first time, the copy of it will be saved as 
+        original mesh. If spline is already defined and in transformed status, 
+        it applies transformation directly.
+
+        Parameters
+        -----------
+        mesh: MESH_TYPES
+            Mesh used for the FFD
+
+        Returns
+        --------
+        None
+        """
+        if self._mesh is not None:
+            self._logw("Resetting original Mesh, this is not intended "
+            "behavior. Please create a new FFD object when replacing the mesh"
+            ".")
+
+        self._logi("Setting mesh.")
+        self._logi("Mesh Info:")
+        self._logi(
+            "  Vertices: {v}.".format(v=mesh.vertices.shape)
+        )
+        self._logi(
+            "  Bounds: {b}.".format(b=mesh.get_bounds())
+        )
+        self._mesh = mesh.copy() # copy to make sure given mesh stays untouched
+
+        self._scale_mesh_vertices()
+        self.calculated = False
+
+    @property
+    def spline(self):
+        """
+        Returns a copy of current spline. This is to prevent unnecessary
+        tangling
```

Comment:
What do you mean with unnecessary tangling. 
You could actually prevent mesh tangling if you use the jacobian. Maybe a future feature?